storagenode/piecestore/endpoint.go

```diff
@@ -314,7 +334,46 @@ func (endpoint *Endpoint) Upload(stream pb.DRPCPiecestore_UploadStream) (err err
 	largestOrder := pb.Order{}
 	defer commitOrderToStore(ctx, &largestOrder)
 
+	// Measure upload rate and average upload rate
+	uploadSize := int64(0)
+	previousSize := int64(0)
+
+	// sumUploadRate sums all previously sampled upload rate
+	sumUploadRate := float64(0)
+	// averageUploadRate= sum of sampled upload rate / total of samples
+	averageUploadRate := float64(0)
+	// number as of data chunks that have been received so far
+	// first chunk is counted as 0
+	chunkCount := float64(-1)
+
 	for {
+		// Increment counts
+		chunkCount++
+		// Measure upload rate (bytes per second)
+		currentTime := time.Now().UTC()
+		dt := currentTime.Sub(startTime)
+
+		if pieceWriter != nil {
+			// delta uploaded
+			uploadSize = pieceWriter.Size() - previousSize
+			previousSize = pieceWriter.Size()
+		}
+
+		if dt.Seconds() > 0 {
```

Comment:
@thepaul, note, this is not always the case. For cross-platform assume that `time.Now` can at most measure increments of 500us.

Reply:
Okay (and wow!), but (a) I'm pretty sure we have nanoseconds on all the platforms we support for the storagenode, and (b) the main point was that I don't think this check is accomplishing what the author intended.

Reply:
(a) we support Windows and the output for this program https://play.golang.org/p/w-xbKFmQ-D1, on windows is:

```
zero count 9999762

  avg 483µs;  min 36.5µs;  p50 500µs;  max 504µs;
  p90 501µs;  p99 504µs;  p999 504µs;  p9999 504µs;
     36.5µs [  3] █
       50µs [  3] █
      100µs [  2] ▌
      150µs [  0]
      200µs [  0]
      250µs [  0]
      300µs [  0]
      350µs [  2] ▌
      400µs [  3] █
      450µs [105] ███████████████████████████████████
      500µs [119] ████████████████████████████████████████
      550µs [  0]
      600µs [  0]
      650µs [  0]
      700µs [  0]
```

(b) ack.

Reply:
Wow! I thought for sure Go would be using GetSystemTimePreciseAsFileTime where possible. But ok, good to know.

Reply:
Results for using `GetSystemTimePreciseAsFileTime`. 100ns is definitely better, but it's still possible to get zeros.

```
zero count 1385005

  avg 106ns;  min 100ns;  p50 100ns;  max 126µs;
  p90 100ns;  p99 200ns;  p999 1µs;  p9999 3.6µs;
      100ns [8614842] ████████████████████████████████████████
       10µs [     51]
       20µs [     23]
       30µs [     22]
       40µs [     27]
       50µs [     10]
       60µs [      7]
       70µs [      5]
       80µs [      3]
       90µs [      1]
      100µs [      1]
      110µs [      1]
      120µs [      1]
      130µs [      0]
      140µs [      0]
```

Note, it's possible to get zero values on other platforms as well and in containers.

AFAIU, `GetSystemTimePreciseAsFileTime` is not being used, because it has a high overhead. https://github.com/golang/go/issues/8687#issuecomment-291570953

Reply:
Note, the call overhead is ~88ns.

Reply:
Thanks. An explicit check for 0 value is added.